### PR TITLE
Shutdown

### PR DESCRIPTION
**Pull request by abhineet-gupta**
## Description and Purpose
Add more shutdown triggers to ROSCO.

## Type of change
What types of change is it?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [x] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

TODO Items General:
- [x] Add example/test for new feature
- [x] Update registry
- [x] Run testing

TODO Items API Change:
- [ ] Update docs with API change
- [x] Run update_rosco_discons.py in Test_Cases/
- [ ] Update DISCON schema

## Examples/Testing, if applicable
Add example 30_shutdown

### CODE REVIEW

File: Examples/example_inputs/minimal_DISCON.IN [hunk 1]
```diff
@@ -92,8 +92,20 @@ Extra lines that ROSCO doesn't care about anymore
...
 
 !------- SHUTDOWN -----------------------------------------------------------
-0.436300000000      ! SD_MaxPit         - Maximum blade pitch angle to initiate shutdown, [rad]
-0.418880000000      ! SD_CornerFreq     - Cutoff Frequency for first order low-pass filter for blade pitch angle, [rad/s]
+0                 ! SD_EnablePitch          - Shutdown when collective blade pitch exceeds a threshold, [-]
+0                 ! SD_EnableYawError        - Shutdown when yaw error exceeds a threshold, [-]
```
Comment (dzalkind) [inline Examples/example_inputs/minimal_DISCON.IN:96]: Are these required inputs?  If not, can they be removed here and run without error?

File: rosco/controller/src/ControllerBlocks.f90 [hunk 2]
```diff
@@ -137,9 +137,9 @@ SUBROUTINE ComputeVariablesSetpoints(CntrPar, LocalVar, objInst, DebugVar, ErrVa
         ENDIF
 
         ! Force zero torque in shutdown mode
-        IF (LocalVar%SD) THEN
-            LocalVar%VS_RefSpd = CntrPar%VS_MinOMSpd
-        ENDIF
+        !IF (LocalVar%SD_Trigger) THEN
```
Comment (dzalkind) [inline rosco/controller/src/ControllerBlocks.f90:140]: Can we remove this?

File: rosco/controller/src/Controllers.f90 [hunk 3]
```diff
@@ -139,7 +134,20 @@ SUBROUTINE PitchControl(avrSWAP, CntrPar, LocalVar, objInst, DebugVar, ErrVar)
         IF (CntrPar%AWC_Mode > 0) THEN
             CALL ActiveWakeControl(CntrPar, LocalVar, DebugVar)
         ENDIF
-
+             
+        ! Shutdown
+        IF (LocalVar%SD_Trigger == 0) THEN
+            LocalVar%PitCom_SD = LocalVar%PitCom
+        ELSE
+            IF (CntrPar%SD_Method == 1) THEN    !Only SD_Method==1 supported for now
```
Comment (dzalkind) [inline rosco/controller/src/Controllers.f90:142]: A comment could help here, like "Pitch to feather at SD_MaxPitchRate".

Note that in some instances (like a downwind rotor), we may want to pitch to a stall angle.

File: rosco/controller/src/DISCON.F90 [hunk 4]
```diff
@@ -100,6 +100,7 @@ SUBROUTINE DISCON(avrSWAP, aviFAIL, accINFILE, avcOUTNAME, avcMSG) BIND (C, NAME
         CALL UpdateZeroMQ(LocalVar, CntrPar, ErrVar)
     ENDIF
     
+    CALL Shutdown(LocalVar, CntrPar, objInst,ErrVar)
```
Comment (dzalkind) [inline rosco/controller/src/DISCON.F90:103]: Should there be an `IF (SD_Mode > 0)` here?

File: rosco/controller/src/ReadSetParameters.f90 [hunk 5]
```diff
@@ -514,8 +514,20 @@ SUBROUTINE ReadControlParameterFileSub(CntrPar, LocalVar, accINFILE, accINFILE_s
...
+        CALL ParseInput(FileLines,  'SD_EnableYawError',     CntrPar%SD_EnableYawError,      accINFILE(1),   ErrVar, CntrPar%SD_Mode == 0, UnEc)
+        CALL ParseInput(FileLines,  'SD_EnableGenSpeed',     CntrPar%SD_EnableGenSpeed,      accINFILE(1),   ErrVar, CntrPar%SD_Mode == 0, UnEc)
+        CALL ParseInput(FileLines,  'SD_EnableTime',         CntrPar%SD_EnableTime,      accINFILE(1),   ErrVar, CntrPar%SD_Mode == 0, UnEc)
+        CALL ParseInput(FileLines,  'SD_MaxPit',             CntrPar%SD_MaxPit,      accINFILE(1),   ErrVar, CntrPar%SD_Mode == 0, UnEc)
+        CALL ParseInput(FileLines,  'SD_PitchCornerFreq',    CntrPar%SD_PitchCornerFreq,      accINFILE(1),   ErrVar, CntrPar%SD_Mode == 0, UnEc)
+        CALL ParseInput(FileLines,  'SD_MaxYawError',        CntrPar%SD_MaxYawError,      accINFILE(1),   ErrVar, CntrPar%SD_Mode == 0, UnEc)
+        CALL ParseInput(FileLines,  'SD_YawErrorCornerFreq', CntrPar%SD_YawErrorCornerFreq,      accINFILE(1),   ErrVar, CntrPar%SD_Mode == 0, UnEc)
+        CALL ParseInput(FileLines,  'SD_MaxGenSpd',          CntrPar%SD_MaxGenSpd,      accINFILE(1),   ErrVar, CntrPar%SD_Mode == 0, UnEc)
```
Comment (dzalkind) [inline rosco/controller/src/ReadSetParameters.f90:525]: Do we want some of these to allow defaults if the enable is 0?

File: rosco/toolbox/controller.py [hunk 6]
```diff
@@ -350,10 +363,10 @@ def tune_controller(self, turbine):
         self.pc_minspd = self.vs_minspd
 
         # max pitch angle for shutdown
-        if self.sd_maxpit:
-            self.sd_maxpit = self.sd_maxpit
-        else:
-            self.sd_maxpit = pitch_op[-1]
+        # if self.sd_maxpit:
```
Comment (dzalkind) [inline rosco/toolbox/controller.py:366]: Clean up?

File: Examples/example_inputs/minimal_DISCON.IN (hunk 1 above)
Comment (abhineet-gupta) [reply Examples/example_inputs/minimal_DISCON.IN:96]: I want these to be optional inputs needed only when SD_Mode = 1. I'll check this and update.

File: rosco/controller/src/ControllerBlocks.f90 (hunk 2 above)
Comment (abhineet-gupta) [reply rosco/controller/src/ControllerBlocks.f90:140]: Yes.

File: rosco/controller/src/ControllerBlocks.f90 [hunk 7]
```diff
@@ -512,47 +512,58 @@ REAL(DbKi) FUNCTION PitchSaturation(LocalVar, CntrPar, objInst, DebugVar, ErrVar
...
+            LocalVar%SD_Trigger = 0
         ENDIF
+        
 
+        ! Filter pitch signal
+        SD_BlPitchF = LPFilter(LocalVar%PC_PitComT, LocalVar%DT, CntrPar%SD_PitchCornerFreq, LocalVar%FP,LocalVar%iStatus, LocalVar%restart, objInst%instLPF)
+        ! Filter yaw error signal (NacVane)
+        SD_NacVaneF = LPFilter(LocalVar%NacVane, LocalVar%DT, CntrPar%SD_YawErrorCornerFreq, LocalVar%FP,LocalVar%iStatus, LocalVar%restart, objInst%instLPF)
+        ! Filter yaw generator speed
+        SD_GenSpeedF = LPFilter(LocalVar%Genspeed, LocalVar%DT, CntrPar%SD_GenSpdCornerFreq, LocalVar%FP,LocalVar%iStatus, LocalVar%restart, objInst%instLPF)
```
[the remark this answers is not in the document]
Comment (abhineet-gupta) [reply rosco/controller/src/ControllerBlocks.f90:542]: Yes, I will make that change.

File: rosco/controller/src/DISCON.F90 (hunk 4 above)
Comment (abhineet-gupta) [reply rosco/controller/src/DISCON.F90:103]: Yes, that makes sense. Making the change.

File: rosco/controller/src/ReadSetParameters.f90 (hunk 5 above)
Comment (abhineet-gupta) [reply rosco/controller/src/ReadSetParameters.f90:525]: Yes, I had added defaults for the toolbox but would be a good idea to set defaults on the controller too. Not sure where controller defaults go but will figure it out.

File: rosco/toolbox/controller.py [hunk 8]
```diff
@@ -98,6 +97,19 @@ def __init__(self, controller_params):
         self.Ki_ipc2p           = controller_params['IPC_Kp2p']
         self.IPC_Vramp          = controller_params['IPC_Vramp']
         self.ZMQ_UpdatePeriod   = controller_params['ZMQ_UpdatePeriod']
+        ## shutdownparameters
+        self.sd_enablepitch = controller_params['sd_enablepitch']
```
[the remark this answers is not in the document]
Comment (abhineet-gupta) [reply rosco/toolbox/controller.py:101]: Ok. Will fix.

File: Examples/example_inputs/minimal_DISCON.IN (hunk 1 above)
Comment (abhineet-gupta) [reply Examples/example_inputs/minimal_DISCON.IN]: Resolved in f674adc

File: rosco/toolbox/controller.py (hunk 6 above)
Comment (abhineet-gupta) [reply rosco/toolbox/controller.py]: Resolved in 809e632

File: rosco/controller/src/ControllerBlocks.f90 [hunk 9]
```diff
@@ -512,47 +512,58 @@ REAL(DbKi) FUNCTION PitchSaturation(LocalVar, CntrPar, objInst, DebugVar, ErrVar
...
         IF (LocalVar%iStatus == 0) THEN
-            LocalVar%SD = .FALSE.
+            LocalVar%SD_Trigger = 0
         ENDIF
+        
 
+        ! Filter pitch signal
+        SD_BlPitchF = LPFilter(LocalVar%PC_PitComT, LocalVar%DT, CntrPar%SD_PitchCornerFreq, LocalVar%FP,LocalVar%iStatus, LocalVar%restart, objInst%instLPF)
+        ! Filter yaw error signal (NacVane)
+        SD_NacVaneF = LPFilter(LocalVar%NacVane, LocalVar%DT, CntrPar%SD_YawErrorCornerFreq, LocalVar%FP,LocalVar%iStatus, LocalVar%restart, objInst%instLPF)
```
[the remark this answers is not in the document]
Comment (abhineet-gupta) [reply rosco/controller/src/ControllerBlocks.f90]: Resolved in 809e632

File: rosco/controller/src/ControllerBlocks.f90 (hunk 7 above)
[the remark this answers is not in the document]
Comment (abhineet-gupta) [reply rosco/controller/src/ControllerBlocks.f90]: Resolved in b57df54

File: rosco/controller/src/Controllers.f90 (hunk 3 above)
Comment (abhineet-gupta) [reply rosco/controller/src/Controllers.f90:143]: Resolved in 96f1f545

File: rosco/controller/src/DISCON.F90 (hunk 4 above)
Comment (abhineet-gupta) [reply rosco/controller/src/DISCON.F90]: resolved in 96f1f545

File: rosco/controller/src/ReadSetParameters.f90 (hunk 5 above)
Comment (abhineet-gupta) [reply rosco/controller/src/ReadSetParameters.f90:525]: Resolved in cca0117

File: rosco/toolbox/controller.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (abhineet-gupta) [reply rosco/toolbox/controller.py]: Resolved in cca0117

**Comment by abhineet-gupta** [comment]
Not sure why tests are failing. `test_examples.py` runs successfully both on macos and linux.

**Comment by abhineet-gupta** [comment]
Question: 
Running [update_rosco_discons.py](https://github.com/NREL/ROSCO/blob/main/Examples/Test_Cases/update_rosco_discons.py) modifies all the DISCON.IN files in `Test_Cases` as expected but also changes `VS_Rgn2K`. Is this expected?
Note that the same happens on the dev branch and `update_rosco_discons.py` gets executed during regression testing [here](https://github.com/NREL/ROSCO/blob/b1a692cff870f18f860cceb20a9a7c7479298ab8/rosco/test/test_examples.py#L35).

**Comment by dzalkind** [comment]
Thanks @abhineet-gupta for your contribution here!  I think this will help in our DLC analysis and provides a good structure for future methods as well.  

I'll merge this once it passes the tests.  A merge conflict broke one of the tests.

**Pull request merged**